src/nvim/message.c

```diff
@@ -1034,16 +1053,50 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
 
   if (msg_hist_off || msg_silent != 0) {
     hl_msg_free(msg);
-    return;
+    return 0;
   }
 
-  // Allocate an entry and add the message at the end of the history.
-  MessageHistoryEntry *entry = xmalloc(sizeof(MessageHistoryEntry));
+  bool progress_message_found = false;
+
+  MessageHistoryEntry *entry = msg_find_by_id(msg_id);
+  if (entry) {
+    progress_message_found = true;
+    // detach the node if found
```

Comment:
When a message gets updated should we drop the message to end of the history? or leave it where it is?

Reply:
Why bother manipulating the message history at all? Can we not just add a new entry for each message (if `history == true`, which it arguably shouldn't be for "progress" messages at least).

Being able to replace a message is about replacing it in the UI right? How the history looks like seems insignificant, and may even be unwanted?

Reply:
Yes I'm also not sure this needs to leak into the C side in the extent it is currently proposed. I would think just assigning a message ID, which is then passed along the the `msg_show` event, is sufficient. A UI can then replace the message text in case the message is still visible somewhere.

What is the purpose of the title, status and percentage fields? Can the caller not represent all of those by modifying the message chunks?

Reply:
> Can the caller not represent all of those by modifying the message chunks?

Technically yes. With kind being able to be anything client can even just drop e json blob in the msg and ui can handle it. But since we are adding explicit support for progress message internally I think it makes sense to keep them as separate fields. When we through autocmd events we can provide them separately then receiver can deal with it how it wants.